### PR TITLE
[V1][Neuron] Neuron chunked prefill V1 impl

### PR DESCRIPTION
**Pull request by elaineyz**
## Purpose

This is the first PR to add support for NxD Inference on vLLM V1 architecture. 

This PR offers a native implementation of Chunked Prefill on Neuron as described in **Change 1** of RFC https://github.com/vllm-project/vllm/issues/21082. 

## Test Plan

An E2E test for chunked prefill is added. Note that the previous E2E tests meant to run on vLLM V0 engine have been removed from the Neuron buildkite script in anticipation for V0 deprecation.

## Test Result

Manually ran test locally on a trainium instance and the test passed

### CODE REVIEW

File: examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        "is_block_kv_layout": True,
+        "sequence_parallel_enabled": True,
+        "on_device_sampling_config": OnDeviceSamplingConfig(),
+        "chunked_prefill_config": {
+            "max_num_seqs": 8,
+            "kernel_q_tile_size": 128,
+            "kernel_kv_tile_size": 4096,
+        },
+        "skip_warmup": True,
+        "save_sharded_checkpoint": True,
```
Comment (aarondou) [inline examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py]: would remove this as it is optional for the purpose of the example. it is off by default

File: examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    override_neuron_config={
+        "is_block_kv_layout": True,
+        "sequence_parallel_enabled": True,
+        "on_device_sampling_config": OnDeviceSamplingConfig(),
+        "chunked_prefill_config": {
+            "max_num_seqs": 8,
+            "kernel_q_tile_size": 128,
+            "kernel_kv_tile_size": 4096,
+        },
+        "skip_warmup": True,
```
Comment (aarondou) [inline examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py:70]: could we test if this can be removed?

File: examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""
+This example is used to illustrate the usage when chunked prefill is enabled.
+To run it, you need to set DISABLE_NEURON_CUSTOM_SCHEDULER=1 when with Neuron
+plugin installed.
```
Comment (aarondou) [inline examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py]: let's add a comment to instruct the chunked prefill doesn't support LNC2, so can only run on trn1 or trn2 with LNC1

File: tests/neuron/2_core/test_chunked_prefill.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+         "blaze like meteors and be gay, Rage, rage against the dying of the "
+         "light. And you, my father, there on the sad height, Curse, bless, me "
+         "now with your fierce tears, I pray. Do not go gentle into that good "
+         "night. Rage, rage against the dying of the light."),
+    ]
+    sampling_params = SamplingParams(max_tokens=30, top_k=1)
+
+    outputs = llm.generate(prompts, sampling_params)
+
+    expected_outputs = [
```
Comment (aarondou) [inline tests/neuron/2_core/test_chunked_prefill.py:63]: are the outputs deterministic? doesn't we need to set the random seed?

File: vllm/platforms/neuron.py [hunk 5]
```diff
@@ -45,17 +45,23 @@ def is_async_output_supported(cls, enforce_eager: Optional[bool]) -> bool:
...
+                # TODO: validate the config. for example, some configs
+                # must (e.g., block_size, enable_chunked_prefill,
+                # etc.) be set or provided with default values.
+            else:
+                parallel_config.worker_cls = \
+                    "vllm.worker.neuron_worker.NeuronWorker"
+                if vllm_config.cache_config and vllm_config.model_config:
+                    # neuron needs block_size = max_model_len
+                    vllm_config.cache_config.block_size = \
+                        vllm_config.model_config.max_model_len  # type: ignore
```
Comment (aarondou) [inline vllm/platforms/neuron.py:60]: this branch will need to be removed (in the PR or follow up) given that v0 is code paths will be gone

File: vllm/v1/worker/neuron_worker.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            vllm_config, self.device)
+
+    def init_device(self) -> None:
+        self.init_distributed_environment()
+
+        # Set random seed.
+        set_random_seed(self.model_config.seed)
+
+    def determine_available_memory(self):
+        # TODO: implement this
```
Comment (aarondou) [inline vllm/v1/worker/neuron_worker.py]: maybe change this to a note that this doesn't matter for nxdi, but needs to be set.
we will need to properly implement this in vllm native.

File: vllm/v1/worker/neuronx_distributed_model_loader.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+    def _compile_and_load_model(self, model_path: str, neuronx_model_cls,
+                                config, compiled_path: str):
+        self.model = neuronx_model_cls(model_path, config)
+        self.model.compile(compiled_path)
+        self.model.load(compiled_path)
+
+
+class NeuronCausalLM(NeuronModelBase):
+
+    def forward(self, input_ids, input_block_ids, **kwargs):
```
Comment (aarondou) [inline vllm/v1/worker/neuronx_distributed_model_loader.py:217]: would fix

File: vllm/v1/worker/neuronx_distributed_model_loader.py [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+                                         config, compiled_model_path)
+
+
+def _get_model_configs(config: PretrainedConfig) -> str:
+    archs = getattr(config, "architectures", [])
+    num_key_value_heads = getattr(config, "num_key_value_heads", None)
+    head_dim = getattr(config, "head_dim", None)
+    if not archs or not num_key_value_heads or not head_dim:
+        raise ValueError("Missing required fields in the model config.")
+    return archs[0], int(num_key_value_heads), int(head_dim)
```
Comment (aarondou) [inline vllm/v1/worker/neuronx_distributed_model_loader.py:276]: would fix

File: vllm/v1/worker/neuronx_distributed_model_loader.py [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+        lora_serving_config)
+
+    neuron_config = _get_neuron_config_after_override(
+        default_neuron_config_args, model_config.override_neuron_config)
+
+    neuron_config = _validate_neuron_config(cache_config, scheduler_config,
+                                            neuron_config)
+
+    override_neuron_config = model_config.override_neuron_config
+    architecture, num_key_value_heads, head_dim = _get_model_configs(
```
Comment (aarondou) [inline vllm/v1/worker/neuronx_distributed_model_loader.py:321]: would fix

File: vllm/v1/worker/neuronx_distributed_model_loader.py [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+        model_config.hf_config)
+    model.load_weights(model_name_or_path=model_config.model,
+                       architecture=architecture,
```
Comment (aarondou) [inline vllm/v1/worker/neuronx_distributed_model_loader.py:324]: would fix

File: vllm/v1/worker/neuronx_distributed_model_loader.py [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+                       neuron_config=neuron_config,
+                       override_neuron_config=override_neuron_config)
+    model.neuron_config = model.model.config.neuron_config
+    model.architecture = architecture
+    model.num_key_value_heads = num_key_value_heads
+    model.head_dim = head_dim
+    return model.eval()
```
Comment (aarondou) [inline vllm/v1/worker/neuronx_distributed_model_loader.py:331]: would fix all above

File: vllm/v1/worker/neuronx_distributed_model_runner.py [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+            # Return empty ModelRunnerOutput if there's no work to do.
+            return EMPTY_MODEL_RUNNER_OUTPUT
+
+        # _prepare_model_input converts the scheduler output to
+        # ModelInputForNeuron
+        model_input = self._prepare_model_input(scheduler_output)
+        logger.debug("model_input: %s", model_input)
+
+        is_mllama = self.model.architecture == "MllamaForConditionalGeneration"
+
```
Comment (aarondou) [inline vllm/v1/worker/neuronx_distributed_model_runner.py:164]: would fix this the and the one above

File: vllm/v1/worker/neuronx_distributed_model_runner.py [hunk 13]
```diff
@@ -0,0 +1,601 @@
...
+                num_kv_heads=self.model.num_key_value_heads,
+                head_size=self.model.head_dim,
+                # TODO: take the following from the model config
+                dtype=torch.bfloat16,
+                use_mla=False,
+                sliding_window=None,
+            )
+        }
+
+    def _update_states(self, scheduler_output: "SchedulerOutput") -> bool:
```
Comment (aarondou) [inline vllm/v1/worker/neuronx_distributed_model_runner.py:249]: would add a note that this is identical to the GPU (TPU as well?).
nonetheless, feel this can be potentially put into the base model runner.

**Comment by aarondou** [comment]
Publish comments

File: examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py (hunk 1 above)
Comment (elaineyz) [reply examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py]: ack

File: examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py (hunk 2 above)
Comment (elaineyz) [reply examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py:70]: yes, inference still works correctly with warmup enabled, but I got very verbose error logs and warmup took 7 minutes. Thus keeping this line for now.

File: examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py (hunk 3 above)
Comment (elaineyz) [reply examples/offline_inference/neuron-v1/neuronx_distributed_chunked_prefill.py]: good catch. I added `"logical_nc_config": 1,`  directly to the override_neuron_config with a comment

File: tests/neuron/2_core/test_chunked_prefill.py (hunk 4 above)
Comment (elaineyz) [reply tests/neuron/2_core/test_chunked_prefill.py:63]: yes the outputs should be deterministic since we're using greedy sampling

File: vllm/platforms/neuron.py (hunk 5 above)
Comment (elaineyz) [reply vllm/platforms/neuron.py:60]: Prefer to leave v0 deletion to a separate PR for simplicity

File: vllm/v1/worker/neuron_worker.py (hunk 6 above)
Comment (elaineyz) [reply vllm/v1/worker/neuron_worker.py]: ack

File: vllm/v1/worker/neuronx_distributed_model_runner.py (hunk 13 above)
Comment (elaineyz) [reply vllm/v1/worker/neuronx_distributed_model_runner.py:249]: makes sense, will add a note

File: vllm/v1/worker/neuronx_distributed_model_loader.py (hunk 9 above)
Comment (elaineyz) [reply vllm/v1/worker/neuronx_distributed_model_loader.py:321]: fixed all mypy warnings/errors
